ci/test/skip

```diff
@@ -0,0 +1,74 @@
+#!/bin/bash
+set -ex
+# shellcheck source=./ci/test/common
+source ./common
+init_test_repo
+cd "$test_repo_path"
+
+cat <<EOF > main.workflow
+workflow "sample"{
+    resolves = ["curl urls"]
+}
+
+action "curl github" {
+  uses = "actions/bin/curl@master"
+  args = ["github.com"]
+}
+
+action "curl google" {
+  uses = "actions/bin/curl@master"
+  args = ["google.com"]
+}
+
+action "curl facebook" {
+  uses = "actions/bin/curl@master"
+  args = ["facebook.com"]
+}
+
+action "curl gitlab" {
+  uses = "actions/bin/curl@master"
+  args = ["gitlab.com"]
+}
+
+action "curl bitbucket" {
+  uses = "actions/bin/curl@master"
+  args = ["bitbucket.org"]
+}
+
+action "curl medium" {
+  uses = "actions/bin/curl@master"
+  args = ["medium.com"]
+}
+
+action "curl twitter" {
+  uses = "actions/bin/curl@master"
+  args = ["twitter.com"]
+}
+
+action "curl youtube" {
+  uses = "actions/bin/curl@master"
+  args = ["youtube.com"]
+}
+
+action "curl do" {
+  uses = "actions/bin/curl@master"
+  args = ["digitalocean.com"]
+}
+
+action "curl urls" {
+    uses = "actions/bin/curl@master"
+    args = ["devfolio.co"]
+    needs = [
+        "curl google",
+        "curl facebook",
+        "curl bitbucket",
+        "curl medium",
+        "curl youtube",
+        "curl twitter",
+    ]
+}
+EOF
+
+! popper run --skip 'curl xyz'
+! popper run --skip 'curl google' --skip 'curl facebook'
```

Comment:
Sorry @jayjeetatgithub, I think I did not describe the definition of a "gap" properly. A gap is present when an action depends on a previous stage (one or more actions), but no actions for that stage are available in the workflow. So for example, for this workflow:

```hcl
workflow "wf" {
  resolves = "d"
}

action "a" {
  uses = "sh"
}

action "b" {
  needs = "a"
  uses = "sh"
}

action "c" {
  needs = "a"
  uses = "sh"
}

action "d" {
  needs = ["b", "c"]
  uses = "sh"
}
```

If we say `popper run --skip b` the workflow should run OK. In this case, we remove `b` from the workflow, and remove it as a dependency from `d` (and as a `next` item in `a`'s list of forward edges). This also should work if we do `popper run --skip c`.

If we skip `b` and `c` though, an error will be thrown because `a` cannot reach `d` (because there is a gap).

Reply:
Thanks a lot. @ivotron.

Reply:
@ivotron Can we place the `remove_skip_actions()` just before `complete_graph()` ? In this way we can leverage `complete_graph()`'s capability to find out unreachable actions. Does this sound correct ?

Reply:
the goal of the `complete_graph` function is to add forward-going edges and the root. We can add another `check_for_unreachable_nodes` function, or something alike. The goal being to have a function do a single thing to improve readability and maintainability (see [my comment](https://github.com/systemslab/popper/pull/650#discussion_r286936204) on #650).

Reply:
Thanks @ivotron.

Reply:
@ivotron  Can the order of calling the functions be like this ? 
```
self.remove_skip_actions()
self.complete_graph()
self.check_for_unreachable_nodes()`
```
In this way, we first remove the actions to be skipped, then complete the graph, and finally check for gaps? Otherwise if `remove_skip_actions()` is after `complete_graph()`, we have to again manipulate the graph.

Reply:
Sorry @JayjeetAtGithub, I should have expressed better what I have in mind.

The idea of running `complete_graph` first is to have an in-memory representation of what the `.workflow` file contains. It would simplify the code if we can have the `Workflow` object be immutable after it is constructed. Then we can apply distinct operations to the graph that return new workflow objects, for example:

 1. filter all actions but one. This is what we do when we are given an action argument.
 2. get the path from the root to a particular action (used when `--with-dependencies` is given).
 3. remove nodes (when `--skip` is given)

Each of the above can be implemented in such a way that they return a new workflow object that is the result of applying the operation to the complete workflow. This would be a functional-ish approach to manipulating the graph, as a way of making it more readable/maintainable. Then we can add other functions that do things like checking for unreachable nodes, node generators (to support distinct ways of traversing the graph), etc. 

The idea is to make the `Workflow` object a generic abstraction that can potentially be reused in multiple scenarios. For example, running the workflow multiple times (but with distinct objectives each time) in a single `popper run` invocation; e.g. run just an action first, then a particular path, then the entire workflow, etc. If we write things in such a way that the original workflow data is lost (e.g. by permanently removing nodes out of the graph), then we won't be able to implement this. Another scenario would be when popper is used as a library (e.g. in a web service), where it would be undesirable to have to read a `.workflow` file multiple times given that the file has not been modified.

So in summary, we can refactor the code so that:

  * the `Workflow`'s constructor only builds the graph, as given in the `.workflow` file.
  * the above means we need to refactor `complete_graph` so that it only completes the graph and does not do any check for unreachable nodes. 
  * add operations for the three things mentioned above
  * add a `check_for_unreachable_nodes`
  * any other thing I'm missing

what do you think?

Reply:
@ivotron Sorry for giving you all the trouble. I now understand what needs to be done completely.
I would merge #621, #622 in this pr. Thanks a lot !

Reply:
@ivotron Does removal of `a` block in the above workflow, leads to unreachable node `b`, `c`,`d` ?

Reply:
@ivotron Also, currently we show a warning for unreachable actions. So would that  case be changed to a fail, and occur through `check_unreachable_nodes()`?

Reply:
@ivotron I have done some refactoring according to your review, and solved issues #621, #626, #622 through this pr. Please review. Thanks.

Reply:
awesome, thanks a lot!! I've left a review

Reply:
> @ivotron Also would the new functions like `remove_skip_actions()`, `check_unreachable_nodes()` etc that would be implemented be part of `Workflow` class, or seperate functions ? In the first case, the function call would look something like `new_wf = old_wf.remove_skip_actions()` and in the second one it will be like `new_wf = remove_skip_actions(old_wf)`.
> 
> Which approach would be better ?

@JayjeetAtGithub sorry, I missed this question. I'd vote for the latter (i.e. as static members of the class).